File: Lib/Kvaser/Canlib/Samples/NET/vs2010/CPlusPlusEvent/Transmitter.cpp
#include "StdAfx.h"
#include "Transmitter.h"

using namespace CPlusPlusEvent;
using namespace canlibCLSNET;

Transmitter::Transmitter(void)
{
	InitializeComponent();
	//
	//TODO: Add the constructor code here
	//

	// restrict control entry
	StandardRB->Checked = true;

	canlibCLSNET::Canlib::canStatus status;

	// Open the channel (The flags will allow a virtual channel to be used)
	channelHandle = canlibCLSNET::Canlib::canOpenChannel(1, canlibCLSNET::Canlib::canOPEN_ACCEPT_VIRTUAL);
	DisplayError(((Canlib::canStatus)channelHandle), "canOpenChannel");

	// Set the bit params (In this case 250 kbits/sec)
	status = canlibCLSNET::Canlib::canSetBusParams(channelHandle, canlibCLSNET::Canlib::canBITRATE_250K, 0, 0, 0, 0, 0);
	DisplayError(status, "canSetBusParams");

	// Make the channel active
	status = canlibCLSNET::Canlib::canBusOn(channelHandle);
	DisplayError(status, "canBusOn");
}


void Transmitter::DisplayError(canlibCLSNET::Canlib::canStatus errorCode, String^ msgTitle)
{
	String^ msgDescr = " ";
	if (errorCode < canlibCLSNET::Canlib::canStatus::canOK)
	{
		if (canlibCLSNET::Canlib::canGetErrorText(errorCode, msgDescr) != 
			canlibCLSNET::Canlib::canStatus::canOK)
		{
			msgDescr = "Unknown error";
		}
		MessageBox::Show(msgTitle + " returned error code " + ((int)errorCode).ToString() + " - " + msgDescr, 
			             msgTitle, 
						 MessageBoxButtons::OK, 
						 MessageBoxIcon::Exclamation);
	}
}

System::Void Transmitter::Transmitter_FormClosing(System::Object^  sender, System::Windows::Forms::FormClosingEventArgs^  e)
{
	Canlib::canStatus status;
	
	// Make the channel inactive
	status = canlibCLSNET::Canlib::canBusOff(channelHandle);
	DisplayError(status, "canBusOff");

	status = canlibCLSNET::Canlib::canClose(channelHandle);
	DisplayError(status, "canClose");
}

System::Void Transmitter::SendButton_Click(System::Object^  sender, System::EventArgs^  e)
{
	Int32 id = 0;
	array<unsigned char>^ data = gcnew array<unsigned char>(8);
	int dlc = 0;
	Int32 flag = 0;

	id = Convert::ToInt32(CanIdTB->Text);
	dlc = Convert::ToInt32(DlcTB->Text);
	data[0] = (unsigned char) Convert::ToInt16(DataTB1->Text);
	data[1] = (unsigned char) Convert::ToInt16(DataTB2->Text);
	data[2] = (unsigned char) Convert::ToInt16(DataTB3->Text);
	data[3] = (unsigned char) Convert::ToInt16(DataTB4->Text);
	data[4] = (unsigned char) Convert::ToInt16(DataTB5->Text);
	data[5] = (unsigned char) Convert::ToInt16(DataTB6->Text);
	data[6] = (unsigned char) Convert::ToInt16(DataTB7->Text);
	data[7] = (unsigned char) Convert::ToInt16(DataTB8->Text);
	if (StandardRB->Checked) {
		flag = canlibCLSNET::Canlib::canMSG_STD;
	}
	else {
		flag = canlibCLSNET::Canlib::canMSG_EXT;
	}
	Canlib::canStatus status = Canlib::canWrite(channelHandle, id, data, dlc, flag);

}

File: Lib/Kvaser/Canlib/Samples/NET/vs2010/VBEvent/Transmitter.vb
Public Class Transmitter

   Dim ChannelHandle As Integer

   Private Sub DisplayError(ByVal errorCode As canlibCLSNET.Canlib.canStatus, ByVal msgTitle As String)
      Dim msgDescr As String = " "
      Dim status As canlibCLSNET.Canlib.canStatus

      If (errorCode < canlibCLSNET.Canlib.canStatus.canOK) Then
         status = canlibCLSNET.Canlib.canGetErrorText(errorCode, msgDescr)
         If (status <> canlibCLSNET.Canlib.canStatus.canOK) Then
            msgDescr = "Unknown error"
         End If
         MessageBox.Show(msgTitle, msgTitle + " returned error code " + Str(errorCode) _
                         + " - " + msgDescr, _
                         MessageBoxButtons.OK, MessageBoxIcon.Exclamation)
      End If
   End Sub

   Public Sub New()

      ' This call is required by the Windows Form Designer.
      InitializeComponent()

      ' restrict control entry
      StandardRB.Checked = True

      ' Add any initialization after the InitializeComponent() call.
      Dim status As canlibCLSNET.Canlib.canStatus

      'Open the channel (The flags will allow a virtual channel to be used)
      ChannelHandle = canlibCLSNET.Canlib.canOpenChannel(1, canlibCLSNET.Canlib.canOPEN_ACCEPT_VIRTUAL)
      DisplayError(ChannelHandle, "canOpenChannel")

      'Set the bit params (In this case 250 kbits/sec)
      status = canlibCLSNET.Canlib.canSetBusParams(ChannelHandle, canlibCLSNET.Canlib.canBITRATE_250K, 0, 0, 0, 0, 0)
      DisplayError(status, "canSetBusParams")

      'Make the channel active
      status = canlibCLSNET.Canlib.canBusOn(ChannelHandle)
      DisplayError(status, "canBusOn")

   End Sub

   Private Sub SendButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles SendButton.Click
      Dim id As Integer = 0
      Dim data As Byte() = New Byte(7) {}
      Dim dlc As Integer = 0
      Dim flag As Integer = 0
      Dim status As canlibCLSNET.Canlib.canStatus

      id = Val(CanIdTB.Text)
      dlc = Val(DlcTB.Text)
      data(0) = Val(DataTB1.Text)
      data(1) = Val(DataTB2.Text)
      data(2) = Val(DataTB3.Text)
      data(3) = Val(DataTB4.Text)
      data(4) = Val(DataTB5.Text)
      data(5) = Val(DataTB6.Text)
      data(6) = Val(DataTB7.Text)
      data(7) = Val(DataTB8.Text)
      If (StandardRB.Checked) Then
         flag = canlibCLSNET.Canlib.canMSG_STD
      Else
         flag = canlibCLSNET.Canlib.canMSG_EXT
      End If

      status = canlibCLSNET.Canlib.canWrite(ChannelHandle, id, data, dlc, flag)

   End Sub

   Private Sub Transmitter_FormClosing(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
      Dim status As canlibCLSNET.Canlib.canStatus

      'Make the channel inactive
      status = canlibCLSNET.Canlib.canBusOff(ChannelHandle)
      DisplayError(status, "canBusOff")

      status = canlibCLSNET.Canlib.canClose(ChannelHandle)
      DisplayError(status, "canClose")

   End Sub
End Class